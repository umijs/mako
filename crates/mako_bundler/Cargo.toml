[package]
name = "mako_bundler"
version = "0.0.0"
edition = "2021"
repository = "https://github.com/umijs/mako"
license = "MIT"

[dependencies]
maplit = "1.0.2"
tokio = { workspace = true, features = ["full"] }
swc_ecma_ast = { version = "0.102.5", features = ["rkyv-impl"] }
swc_ecma_parser = { version = "0.132.6" }
swc_common = { version = "0.30.5", features = ["concurrent"] }
swc_ecma_visit = { version = "0.88.5" }
swc_ecma_codegen = { version = "0.137.6" }
swc_ecma_transforms = { version = "0.216.4", features = [
  "module",
  "typescript",
  "react",
  "concurrent",
] }
swc_ecma_utils = { version = "0.115.8" }
swc_atoms = { version = "0.4.43" }
relative-path = "1.8.0"
regex = "1"
lazy_static = "1.4.0"
base64 = "0.21.0"
petgraph = "0.6.3"
md5 = "0.7.0"
serde = { version = "1.0.160", features = ["derive"] }
serde_json = "1.0.96"
config = "0.13.3"
nodejs-resolver = "0.0.78"
lightningcss = "1.0.0-alpha.41"
swc_css_ast = "0.136.6"
swc_css_codegen = "0.146.6"
swc_css_parser = "0.145.6"
swc_css_visit = "0.135.6"
rayon = "1.7.0"

[dev-dependencies]
insta = { version = "1.29.0", features = ["yaml"] }
