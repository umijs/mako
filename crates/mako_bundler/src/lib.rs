#![feature(box_patterns)]

use compiler::Compiler;

use crate::plugins::node_polyfill::get_node_builtins;

pub mod build;
pub mod chunk;
pub mod chunk_graph;
pub mod compiler;
pub mod config;
pub mod context;
pub mod generate;
pub mod module;
pub mod module_graph;
pub mod plugin;
pub mod plugins;
pub mod utils;

pub fn run() {
    let args: Vec<String> = std::env::args().collect();
    if args.len() <= 1 {
        panic!("Please specify the root directory of the project");
    }

    // config
    let root = std::env::current_dir()
        .unwrap()
        .join(&args[1])
        .to_string_lossy()
        .to_string();
    let mut config = config::Config::from_literal_str(
        format!(
            r#"
{{
    "externals": {{}},
    "root": "{}",
    "entry": {{ "index": "index.tsx" }}
}}
    "#,
            root
        )
        .as_str(),
    )
    .unwrap();
    config.normalize();

    // TODO: move this to plugin
    // node polyfills
    let builtins = get_node_builtins();
    for name in builtins.iter() {
        config.externals.insert(name.to_string(), name.to_string());
    }

    // compiler_origin::run_compiler(config);
    let mut compiler = Compiler::new(config);
    compiler.run();

    println!("âœ… DONE");
}
