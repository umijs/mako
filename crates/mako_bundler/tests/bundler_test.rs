use std::{collections::HashMap, vec};

use mako_bundler::{
    build::build::BuildParam, compiler::Compiler, config::Config,
    generate::generate::GenerateParam, module::ModuleId,
};

#[test]
#[ignore]
fn normal() {
    let files = HashMap::from([
        (
            "/tmp/entry.js".to_string(),
            r###"
import {fn} from './foo';
console.log(fn());
            "###
            .to_string(),
        ),
        (
            "/tmp/foo.js".to_string(),
            r###"
export function fn() {
    return 123
}
            "###
            .to_string(),
        ),
    ]);
    let (output, _) = test_files(files);
    insta::assert_debug_snapshot!(output);
}

#[test]
#[ignore]
fn multiple_files() {
    let files = HashMap::from([
        (
            "/tmp/entry.js".to_string(),
            r###"
import {three} from './three';
import {one} from './one';
import {two} from './two';
console.log(one());
            "###
            .to_string(),
        ),
        (
            "/tmp/one.js".to_string(),
            r###"
import {two} from './two';
export function one() {
    return two();
}
            "###
            .to_string(),
        ),
        (
            "/tmp/two.js".to_string(),
            r###"
export function two() {
    return 123
}
            "###
            .to_string(),
        ),
        (
            "/tmp/three.js".to_string(),
            r###"
export function three() {
    return 123
}
            "###
            .to_string(),
        ),
    ]);
    let (output, mut compiler) = test_files(files);
    insta::assert_debug_snapshot!(output);
    let orders = compiler.context.module_graph.topo_sort().unwrap();
    assert_eq!(
        &orders,
        &vec![
            ModuleId::new("/tmp/entry.js"),
            ModuleId::new("/tmp/three.js"),
            ModuleId::new("/tmp/one.js"),
            ModuleId::new("/tmp/two.js"),
        ]
    );
    let mut vecs = vec![];
    for module_id in orders {
        let deps = compiler.context.module_graph.get_dependencies(&module_id);
        vecs.push((module_id, deps));
    }
    insta::assert_debug_snapshot!(&vecs);
}

#[test]
#[ignore]
fn replace_env() {
    let files = HashMap::from([
        (
            "/tmp/entry.js".to_string(),
            r###"
import {one} from './one';
if (process.env.NODE_ENV === 'production') {
	console.log(123);
}
            "###
            .to_string(),
        ),
        (
            "/tmp/one.js".to_string(),
            r###"
function foo() {
	if (process.env.NODE_ENV === 'production') {
		console.log(123);
	}
	if (process.env['NODE_ENV'] === 'production') {
		console.log(123);
	}
	const test = process.env['NODE_ENV'];
}
            "###
            .to_string(),
        ),
    ]);
    let (output, _) = test_files(files);
    insta::assert_debug_snapshot!(output);
}

#[allow(clippy::useless_format)]
fn test_files(files: HashMap<String, String>) -> (Vec<String>, Compiler) {
    let mut config = Config::from_literal_str(
        format!(
            r#"
{{
    "entry": {{
        "entry": "/tmp/entry.js"
    }},
    "root": "/tmp"
}}
            "#
        )
        .as_str(),
    )
    .unwrap();
    config.normalize();
    let mut compiler = Compiler::new(config);
    compiler.build(&BuildParam { files: Some(files) });
    let generate_result = compiler.generate(&GenerateParam { write: false });
    let output = generate_result.output_files[0].__output.clone();
    (output, compiler)
}
