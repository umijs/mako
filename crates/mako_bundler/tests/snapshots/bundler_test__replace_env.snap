---
source: crates/mako_bundler/tests/bundler_test.rs
expression: output
---
[
    [
        "\nconst modules = new Map();\nconst g_define = (name, moduleFactory) => {\n    modules.set(name, moduleFactory);\n};\n\nfunction _interop_require_default(obj) {\n    return obj && obj.__esModule && obj['default'] ? obj : { default: obj, ...obj };\n}\n\nfunction _getRequireWildcardCache(nodeInterop) {\n    if (typeof WeakMap !== \"function\") return null;\n    var cacheBabelInterop = new WeakMap();\n    var cacheNodeInterop = new WeakMap();\n    return (_getRequireWildcardCache = function(nodeInterop) {\n        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;\n    })(nodeInterop);\n}\nfunction _interop_require_wildcard(obj, nodeInterop) {\n    if (!nodeInterop && obj && obj.__esModule) {\n        return obj;\n    }\n    if (obj === null || typeof obj !== \"object\" && typeof obj !== \"function\") {\n        return { default: obj };\n    }\n    var cache = _getRequireWildcardCache(nodeInterop);\n    if (cache && cache.has(obj)) {\n        return cache.get(obj);\n    }\n    var newObj = {};\n    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;\n    for (var key in obj) {\n        if (key !== \"default\" && Object.prototype.hasOwnProperty.call(obj, key)) {\n            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;\n            if (desc && (desc.get || desc.set)) {\n                Object.defineProperty(newObj, key, desc);\n            } else {\n                newObj[key] = obj[key];\n            }\n        }\n    }\n    newObj.default = obj;\n    if (cache) {\n        cache.set(obj, newObj);\n    }\n    return newObj;\n}\n\nconst moduleCache = new Map();\nconst requireModule = (name) => {\n    if (moduleCache.has(name)) {\n    return moduleCache.get(name).exports;\n    }\n\n    if (!modules.has(name)) {\n    throw new Error(`Module '${name}' does not exist.`);\n    }\n\n    const moduleFactory = modules.get(name);\n    const module = {\n    exports: {},\n    };\n    moduleCache.set(name, module);\n    moduleFactory(module, module.exports, requireModule);\n    return module.exports;\n};\n        ",
        "g_define(\"<CWD>/tests/fixtures/env/entry.js\", function(module, exports, require) {\n\"use strict\";\nObject.defineProperty(exports, \"__esModule\", {\n    value: true\n});\nvar _one = require(\"<CWD>/tests/fixtures/env/one.js\");\nif (\"production\" === 'production') {\n    console.log(123);\n}\n});",
        "g_define(\"<CWD>/tests/fixtures/env/one.js\", function(module, exports, require) {\n\"use strict\";\nfunction foo() {\n    if (\"production\" === 'production') {\n        console.log(123);\n    }\n    if (\"production\" === 'production') {\n        console.log(123);\n    }\n    const test = \"production\";\n}\n});",
        "\nrequireModule(\"<CWD>/tests/fixtures/env/entry.js\");",
    ],
]
